query-engine/connectors/sql-query-connector/src/filter_conversion.rs

```diff
@@ -57,13 +58,115 @@ impl Alias {
     }
 }
 
+#[derive(Clone, Debug, PartialEq)]
+enum QueryOperation {
+    // Insert, Not used yet
+    Update,
+    Delete,
+    // Select, Not used yet
+}
+
+#[derive(Clone)]
+pub struct ConditionState {
+    reverse: bool,
+    alias: Option<Alias>,
+    tables: Vec<Table<'static>>,
+    operation: Option<QueryOperation>,
+    /// SqlInfo should be used very sparingly otherwise its going to blow up our code with lots
+    /// of checks for different
+    sql_info: Option<SqlInfo>,
+}
+
+impl ConditionState {
+    fn new(alias: Option<Alias>, reverse: bool) -> Self {
+        Self {
+            reverse,
+            alias,
+            tables: Vec::new(),
+            operation: None,
+            sql_info: None,
+        }
+    }
+
+    pub fn for_update_query(tables: impl Into<Vec<Table<'static>>>, sql_info: SqlInfo) -> Self {
+        Self {
+            reverse: false,
+            alias: None,
+            tables: tables.into(),
+            operation: Some(QueryOperation::Update),
+            sql_info: Some(sql_info),
+        }
+    }
+
+    pub fn for_delete_query(tables: impl Into<Vec<Table<'static>>>, sql_info: SqlInfo) -> Self {
+        Self {
+            reverse: false,
+            alias: None,
+            tables: tables.into(),
+            operation: Some(QueryOperation::Delete),
+            sql_info: Some(sql_info),
+        }
+    }
+
+    /// MySql will error if a `Update` or `Delete` query has a subselect
+    /// that references a table that is being updated or deleted
+    /// to get around that, we need to wrap the table in a tmp table name
+    ///
+    /// UPDATE top_table
+    /// SET top_table.field = 2
+    /// WHERE top_table.id IN (SELECT * FROM(SELECT top_table.id
+    ///                     FROM top_table t
+    ///                     INNER JOIN join_table ja
+    ///                     ON t.join_id = ja.id
+    ///                     WHERE ja.id = 12) as tblTmp);
+    fn should_wrap_select_in_tmp_table(&self, table: Table) -> bool {
```

Comment:
Could we try to do this in https://github.com/prisma/quaint/blob/main/src/visitor/mysql.rs as we've been doing so far. I'm not super excited to have the filter function deal with the database type. In the future we should have database-specific query builders, but right now this kind of breaks what the original idea of the filter conversion was. In the Prisma 1 times the code became quite messy and we tried to avoid it in Prisma 2.